snapcraft/internal/build_providers/_lxd/_lxd.py

```diff
@@ -350,6 +367,10 @@ def shell(self) -> None:
         self._run(command=["/bin/bash"])
 
     def _ensure_container_running(self) -> None:
+        # Sanity check - developer error if container not initialized.
```

Comment:
would adding the check only here make it so that you don't need to do it in every method? If not, a decorator might be better

Reply:
I do not believe the check in _ensure_container_running was suficient unfortunately

Reply:
Confirmed.  And tried the decorator method but stumbled onto: https://github.com/python/mypy/issues/5836

Reply:
gotcha, sucks a bit though

Reply:
I agree. I'm going to come back to this when we revisit local/remote provider.  I don't think _container should be optional, and just launched on init.  Everything is dependent on it anyhow... and if not, it's probably a classmethod.